### PR TITLE
Encapsulate HLL logic

### PR DESCRIPTION
**Pull request by kangkaisen**
For https://github.com/apache/incubator-doris/pull/1610

Before I do "avoid serialize and deserialize from Scannode to Aggnode" work, I need to unify Bitmap and HyperLogLog behavior and interface firstly. So I encapsulate the HyperLogLog related logic.

### CODE REVIEW

File: be/src/exec/csv_scan_node.cpp [hunk 1]
```diff
@@ -672,14 +672,14 @@ void CsvScanNode::hll_hash(const char* src, int len, std::string* result) {
     std::string str(src, len);
     if (str != "\\N") {
         uint64_t hash = HashUtil::murmur_hash64A(src, len, HashUtil::MURMUR_SEED);
-        char buf[HllHashFunctions::HLL_INIT_EXPLICT_SET_SIZE];
+        char buf[10];
```
Comment (imay) [inline be/src/exec/csv_scan_node.cpp]: We should use a const or a macro, better not to use a magic number

File: be/src/olap/aggregate_func.h [hunk 2]
```diff
@@ -399,64 +399,43 @@ struct AggregateFuncTraits<OLAP_FIELD_AGGREGATION_REPLACE, OLAP_FIELD_TYPE_CHAR>
...
+
+        // fixme(kks): trick here, need improve
+        if (arena == nullptr) { // for query
+            HyperLogLog src_hll = HyperLogLog(src_slice->data);
+            dst_hll->merge(src_hll);
+        } else {   // for stream load
+            auto* src_hll = reinterpret_cast<HyperLogLog*>(src_slice->data);
+            dst_hll->merge(*src_hll);
+
+            delete src_hll;
```
Comment (imay) [inline be/src/olap/aggregate_func.h]: This is a dangerous to delete input source content.
Better to leave a comment and try to find a better way to handle this

File: be/src/exprs/hll_function.cpp [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+}
+
+StringVal HllFunctions::hll_hash(FunctionContext* ctx, const StringVal& input) {
+    const int HLL_SINGLE_VALUE_SIZE = 10;
+    const int HLL_EMPTY_SIZE = 1;
+    std::string buf;
+    std::unique_ptr<HyperLogLog> hll {new HyperLogLog()};
+    if (!input.is_null) {
+        uint64_t hash_value = HashUtil::murmur_hash64A(input.ptr, input.len, HashUtil::MURMUR_SEED);
+        hll.reset(new HyperLogLog(hash_value));
```
Comment (imay) [inline be/src/exprs/hll_function.cpp]: For this case we create HyperLogLog two times

File: be/src/exprs/hll_function.cpp [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+        buf.resize(HLL_SINGLE_VALUE_SIZE);
+    } else {
+        buf.resize(HLL_EMPTY_SIZE);
+    }
+    hll->serialize((char*)buf.c_str());
```
Comment (imay) [inline be/src/exprs/hll_function.cpp]: why not serialize(std::string*) ?

File: be/src/exprs/hll_hash_function.cpp [hunk 5]
```diff
@@ -15,62 +15,34 @@
...
 
-StringVal HllHashFunctions::create_string_result(doris_udf::FunctionContext* ctx, 
-                                                     const StringVal& val, const bool is_null) {
-    StringVal result;
-    if (is_null) {
-        // HLL_DATA_EMPTY
-        char buf[HLL_EMPTY_SET_SIZE];
-        buf[0] = HLL_DATA_EMPTY;
-        result = AnyValUtil::from_buffer_temp(ctx, buf, sizeof(buf));
+StringVal HllHashFunctions::hll_hash(FunctionContext* ctx, const StringVal& input) {
```
Comment (imay) [inline be/src/exprs/hll_hash_function.cpp]: we have two implementation? [here](https://github.com/apache/incubator-doris/pull/1756/files#diff-964de1ed6e9fc1eacc99c8d4f1a680aaR31)

File: be/src/olap/hll.h [hunk 6]
```diff
@@ -33,14 +33,339 @@ const static int HLL_REGISTERS_COUNT = 16384;
...
+    // change the _type to HLL_DATA_FULL directly has two reasons:
+    // 1. keep behavior consistent with before
+    // 2. make the code logic is simple
+    void update(const uint64_t hash_value) {
+        _type = HLL_DATA_FULL;
+        update_registers(_registers, hash_value);
+    }
+
+    void merge(const HyperLogLog& other) {
+        if (other._type == HLL_DATA_EMPTY) {
```
Comment (imay) [inline be/src/olap/hll.h]: why not switch?

File: be/src/olap/hll.h [hunk 7]
```diff
@@ -33,14 +33,339 @@ const static int HLL_REGISTERS_COUNT = 16384;
...
+            return;
+        }
+
+        // _type must change
+        if (_type == HLL_DATA_EMPTY) {
+            _type = other._type;
```
Comment (imay) [inline be/src/olap/hll.h]: if this->_type is changed, do you want it do following if block? I think there is no need to do following if block, which is a waste of time.

File: be/src/olap/hll.h [hunk 8]
```diff
@@ -33,14 +33,339 @@ const static int HLL_REGISTERS_COUNT = 16384;
...
+
+    // change the _type to HLL_DATA_FULL directly has two reasons:
+    // 1. keep behavior consistent with before
+    // 2. make the code logic is simple
+    void update(const uint64_t hash_value) {
+        _type = HLL_DATA_FULL;
+        update_registers(_registers, hash_value);
+    }
+
+    void merge(const HyperLogLog& other) {
```
Comment (imay) [inline be/src/olap/hll.h]: For large code block, we'd better to write it in cpp/cc files to accelerate our compile time

File: be/src/olap/hll.h [hunk 9]
```diff
@@ -33,14 +33,339 @@ const static int HLL_REGISTERS_COUNT = 16384;
...
+        }
+    }
+
+    static int serialize_full(char* result, char* registers) {
+        result[0] = HLL_DATA_FULL;
+        memcpy(result + 1, registers, HLL_REGISTERS_COUNT);
+        return HLL_COLUMN_DEFAULT_LEN;
+    }
+
+    static int serialize_sparse(char *result, const std::map<int, uint8_t>& index_to_value) {
```
Comment (imay) [inline be/src/olap/hll.h]: can this be private?

**Review by imay — CHANGES_REQUESTED** [review]

File: be/src/olap/hll.h (hunk 9 above)
Comment (kangkaisen) [reply be/src/olap/hll.h]: OK

File: be/src/olap/hll.h (hunk 8 above)
Comment (kangkaisen) [reply be/src/olap/hll.h]: OK

File: be/src/olap/hll.h (hunk 7 above)
Comment (kangkaisen) [reply be/src/olap/hll.h]: Yes. So I directly return in every switch case.

File: be/src/olap/hll.h (hunk 6 above)
Comment (kangkaisen) [reply be/src/olap/hll.h]: I think 4 * 3 switch case is too big.

Use if block one by one, and directly return in every switch case, which like Guard Clauses and seems more readable.

File: be/src/exprs/hll_hash_function.cpp (hunk 5 above)
Comment (kangkaisen) [reply be/src/exprs/hll_hash_function.cpp]: Yes. I have commented. keep HllHashFunctions only for backward compatibility, we will remove the HllHashFunctions class  when doris 0.12 release.

File: be/src/exprs/hll_function.cpp (hunk 4 above)
Comment (kangkaisen) [reply be/src/exprs/hll_function.cpp]: Because slice.data type is char* and many HLL methods use char*.

File: be/src/exprs/hll_function.cpp (hunk 3 above)
Comment (kangkaisen) [reply be/src/exprs/hll_function.cpp]: I will Fix

File: be/src/olap/aggregate_func.h (hunk 2 above)
Comment (kangkaisen) [reply be/src/olap/aggregate_func.h]: I will Fix this issue with object_pool in next PR.

File: be/src/exec/csv_scan_node.cpp (hunk 1 above)
Comment (kangkaisen) [reply be/src/exec/csv_scan_node.cpp]: I see. Use magic number only for simple.
Because I think we could delete this class code in doris 0.12 version.

**Review by imay — APPROVED** [review]
LGTM

**Pull request merged**